### PR TITLE
allow adding gateway endpoints to vpc

### PR DESCRIPTION
**Pull request by tareqhs**
**What this PR does / why we need it**:
Allow adding gateway endpoints to the VPC of an AWS cluster. Example for adding an endpoint for S3:
```yaml
apiVersion: core.gardener.cloud/v1alpha1
kind: Shoot
spec:
  provider:
    type: aws
    infrastructureConfig:
      apiVersion: aws.provider.extensions.gardener.cloud/v1alpha1
      kind: InfrastructureConfig
      networks:
        vpc:
          gatewayEndpoints:
          - "s3"
```

More info about gateway endpoints
https://docs.aws.amazon.com/vpc/latest/userguide/vpce-gateway.html

Relevant terraform resource
https://www.terraform.io/docs/providers/aws/r/vpc_endpoint.html

**Release note**:
```improvement operator
Now it is possible to define VPC Gateway Endpoints for AWS shoots. The new list "gatewayEdnpoints" inside "vpc" sections is a list endpoint service names. Each item is mapped to a vpc gateway endpoint for a service "com.amazonaws.<region>.<service>", region being the shoot region.
```

### CODE REVIEW

File: charts/internal/aws-infra/templates/main.tf [hunk 1]
```diff
@@ -24,6 +24,13 @@ resource "aws_vpc" "vpc" {
 {{ include "aws-infra.common-tags" .Values | indent 2 }}
 }
 
+{{ range $ep := .Values.vpc.gatewayEndpoints }}
+resource "aws_vpc_endpoint" "vpc_gwep_{{ $p }}" {
```
Comment (vpnachev) [inline charts/internal/aws-infra/templates/main.tf]: maybe you want 
```suggestion
resource "aws_vpc_endpoint" "vpc_gwep_{{ $ep }}" {
```
?

Comment (tareqhs) [reply charts/internal/aws-infra/templates/main.tf]: correct, was a typo, thanks!

**Comment by tareqhs** [comment]
> You have to sign the CLA.

done

File: pkg/apis/aws/types_infrastructure.go [hunk 2]
```diff
@@ -87,6 +87,8 @@ type VPC struct {
 	ID *string
 	// CIDR is the VPC CIDR.
 	CIDR *string
+	// GatewayEndpoints service names to configure as gateway endpoints in the VPC.
+	GatewayEndpoints []string
```
Comment (rfranzke) [inline pkg/apis/aws/types_infrastructure.go]: You have to add this to `pkg/apis/aws/v1alpha1/types_infrastructure.go` and run `make generate`
Also, please add `// +optional` and `omitempty` tags there.

File: pkg/apis/aws/validation/infrastructure.go [hunk 3]
```diff
@@ -85,6 +86,16 @@ func ValidateInfrastructureConfig(infra *apisaws.InfrastructureConfig, nodesCIDR
 		allErrs = append(allErrs, field.Required(networksPath.Child("zones"), "must specify at least the networks for one zone"))
 	}
 
+	if len(infra.Networks.VPC.GatewayEndpoints) > 0 {
```
Comment (rfranzke) [inline pkg/apis/aws/validation/infrastructure.go]: Can we have a unit test for this?

Comment (tareqhs) [reply pkg/apis/aws/validation/infrastructure.go]: in progress .. why is the voter "concourse-ci/publish" failing on helm stuff?

Comment (rfranzke) [reply pkg/apis/aws/validation/infrastructure.go]:
```
Executing check-generate
go generate needs to be run:
 M pkg/apis/aws/v1alpha1/zz_generated.conversion.go
 M pkg/apis/aws/zz_generated.deepcopy.go
```

Can you read the logs? They should be public.

Comment (tareqhs) [reply pkg/apis/aws/validation/infrastructure.go]: I've run both ‘make generate‘ and ’go generate’, still getting this failure ..

Comment (ialidzhikov) [reply pkg/apis/aws/validation/infrastructure.go]: Currently `make generate` on your PR adds to following changes:
```
$ git status
On branch gatewayendpoints
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   pkg/apis/aws/v1alpha1/zz_generated.conversion.go
	modified:   pkg/apis/aws/zz_generated.deepcopy.go

no changes added to commit (use "git add" and/or "git commit -a")
```

You need to add them as part of your PR.

Comment (ialidzhikov) [reply pkg/apis/aws/validation/infrastructure.go]: What is the `make generate` output for you?

Comment (tareqhs) [reply pkg/apis/aws/validation/infrastructure.go]:
```text
> Generate
Successfully generated controller registration at ../../example/controller-registration.yaml
Generating deepcopy funcs
Generating defaulters
Generating conversions
E0209 13:23:52.260094    5469 conversion.go:667] Warning: could not find nor generate a final Conversion function for github.com/gardener/gardener-extension-provider-aws/pkg/apis/aws.VPC -> github.com/gardener/gardener-extension-provider-aws/pkg/apis/aws/v1alpha1.VPC
E0209 13:23:52.262090    5469 conversion.go:668]   the following fields need manual conversion:
E0209 13:23:52.267350    5469 conversion.go:670]       - GatewayEndpoints
Generating deepcopy funcs
Generating defaulters
Generating conversions
I0209 13:27:53.975702    8837 main.go:123] parsing go packages in directory .
I0209 13:28:38.693263    8837 main.go:225] using package=.
I0209 13:28:38.920095    8837 main.go:161] written to ../../../../hack/api-reference/api.md
I0209 13:28:39.383321    9530 main.go:123] parsing go packages in directory .
I0209 13:29:35.271869    9530 main.go:225] using package=.
I0209 13:29:35.434090    9530 main.go:161] written to ../../../../hack/api-reference/config.md
```

Comment (vpnachev) [reply pkg/apis/aws/validation/infrastructure.go]: See the comment from Rafael above https://github.com/gardener/gardener-extension-provider-aws/pull/17/files#r374106445. 
You have to extend the versioned API for the VPC also: https://github.com/gardener/gardener-extension-provider-aws/blob/641a4095314420a302d165c24875caebcfc1b66a/pkg/apis/aws/v1alpha1/types_infrastructure.go#L86-L94, too

Comment (tareqhs) [reply pkg/apis/aws/validation/infrastructure.go]: @vpnachev , @ialidzhikov , i've done this already
https://github.com/gardener/gardener-extension-provider-aws/blob/5d2344a44a1103bc7c1f97ae1238e2ddfa146191/pkg/apis/aws/v1alpha1/types_infrastructure.go#L96

File: pkg/apis/aws/types_infrastructure.go (hunk 2 above)
Comment (tareqhs) [reply pkg/apis/aws/types_infrastructure.go]: done, still failing

Comment (vpnachev) [reply pkg/apis/aws/types_infrastructure.go]: you have not run `make generate`

File: pkg/apis/aws/validation/infrastructure.go (hunk 3 above)
Comment (tareqhs) [reply pkg/apis/aws/validation/infrastructure.go]: @vpnachev , @ialidzhikov , done

File: docs/usage-as-end-user.md [hunk 4]
```diff
@@ -65,6 +65,22 @@ For every zone, the AWS extension creates three subnets:
 For every subnet, you have to specify a CIDR range contained in the VPC CIDR specified above, or the VPC CIDR of your already existing VPC.
 You can freely choose these CIDR and it is your responsibility to properly design the network layout to suit your needs.
 
+You can configure [Gateway VPC Endpoints](https://docs.aws.amazon.com/vpc/latest/userguide/vpce-gateway.html) by adding items in the optional list `networks.vpc.gatewayEndpoints`. Each item in the list is used as a service name and a corresponding endpoint is created for it. All created endpoints point to the service within the cluster's region. For example, consider this (partial) shoot config:
```
Comment (vpnachev) [inline docs/usage-as-end-user.md]: Could you summarize it and add it here https://github.com/gardener/gardener-extension-provider-aws/pull/17/files#diff-612a4cfc1d7f00d699d1bd3d3465e1f2R57. as a 1 or 2 lines.
Also, you can adapt the existing example https://github.com/gardener/gardener-extension-provider-aws/pull/17/files#diff-612a4cfc1d7f00d699d1bd3d3465e1f2R37-R39 with commented sample values.

Comment (tareqhs) [reply docs/usage-as-end-user.md]: please review the update

File: docs/usage-as-end-user.md [hunk 5]
```diff
@@ -37,6 +37,10 @@ networks:
   vpc: # specify either 'id' or 'cidr'
   # id: vpc-123456
     cidr: 10.250.0.0/16
+  # you can define custom service names for adding Gateway VPC Endpoints
```
Comment (vpnachev) [inline docs/usage-as-end-user.md]: remove this line
```suggestion
```

File: docs/usage-as-end-user.md [hunk 6]
```diff
@@ -37,6 +37,10 @@ networks:
...
+  # gatewayEndpoints:
+  # - s3
+  #   this translate to service name com.amazonaws.<shoot-region>.s3
```
Comment (vpnachev) [inline docs/usage-as-end-user.md]: remove this line
```suggestion
```
